python/cudf/cudf/tests/test_df_protocol.py

```diff
@@ -0,0 +1,220 @@
+from typing import Any, Tuple
+
+import cupy as cp
+import pandas as pd
+import pytest
+
+import cudf
+from cudf.core.buffer import Buffer
+from cudf.core.column import build_column
+from cudf.core.df_protocol import (
+    _CuDFBuffer,
+    _CuDFColumn,
+    _DtypeKind,
+    _from_dataframe,
+    protocol_dtype_to_cupy_dtype,
+)
+from cudf.testing._utils import assert_eq
+
+DataFrameObject = Any
+
+
+def assert_buffer_equal(buffer_and_dtype: Tuple[_CuDFBuffer, Any], cudfcol):
+    buf, dtype = buffer_and_dtype
+    device_id = cp.asarray(cudfcol.data).device.id
+    assert buf.__dlpack_device__() == (2, device_id)
+    col_from_buf = build_column(
+        Buffer(buf.ptr, buf.bufsize), protocol_dtype_to_cupy_dtype(dtype)
+    )
+    # check that non null values are the equals as nulls are represented
+    # by sentinel values in the buffer.
+    non_null_idxs = cudf.Series(cudfcol) != cudf.NA
+    assert_eq(col_from_buf[non_null_idxs], cudfcol[non_null_idxs])
+
+    if dtype[0] != _DtypeKind.BOOL:
+        array_from_dlpack = cp.fromDlpack(buf.__dlpack__())
+        col_array = cp.asarray(cudfcol.data_array_view)
+        assert_eq(array_from_dlpack.flatten(), col_array.flatten())
+    else:
+        pytest.raises(TypeError, buf.__dlpack__)
+
+
+def assert_column_equal(col: _CuDFColumn, cudfcol):
+    assert col.size == cudfcol.size
+    assert col.offset == 0
+    assert col.null_count == cudfcol.null_count
+    assert col.num_chunks() == 1
+    if col.null_count == 0:
+        pytest.raises(RuntimeError, col._get_validity_buffer)
+        assert col.get_buffers()["validity"] is None
+    else:
+        assert_buffer_equal(
+            col.get_buffers()["validity"],
+            cudfcol._get_mask_as_column().astype(cp.uint8),
+        )
+
+    if col.dtype[0] == _DtypeKind.CATEGORICAL:
+        assert_buffer_equal(col.get_buffers()["data"], cudfcol.codes)
+        assert col.get_buffers()["offsets"] is None
+
+    elif col.dtype[0] == _DtypeKind.STRING:
+        assert_buffer_equal(col.get_buffers()["data"], cudfcol.children[1])
+        assert_buffer_equal(col.get_buffers()["offsets"], cudfcol.children[0])
+
+    else:
+        assert_buffer_equal(col.get_buffers()["data"], cudfcol)
+        assert col.get_buffers()["offsets"] is None
+
+    if col.null_count == 0:
+        assert col.describe_null == (0, None)
+    else:
+        assert col.describe_null == (3, 0)
+
+
+def assert_dataframe_equal(dfo: DataFrameObject, df: cudf.DataFrame):
+    assert dfo.num_columns() == len(df.columns)
+    assert dfo.num_rows() == len(df)
+    assert dfo.num_chunks() == 1
+    assert dfo.column_names() == list(df.columns)
+    for col in df.columns:
+        assert_column_equal(dfo.get_column_by_name(col), df[col]._column)
+
+
+def _test_from_dataframe_equals(dfobj):
```

Comment:
I would use a name that is more similar to the helper function `assert_dataframe_equal`. This isn't a test, it's a helper for a test. I would also make it a public name rather than a private name, I think? Not sure what the testing conventions are here but I think it's appropriate to match `assert_dataframe_equal`.

```suggestion
def assert_from_dataframe_equals(dfobj):
```